cub/block/block_run_length_decode.cuh

```diff
@@ -0,0 +1,435 @@
+/******************************************************************************
+ * Copyright (c) 2011-2021, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include "../config.cuh"
+#include "../util_ptx.cuh"
+#include "../util_type.cuh"
+#include "block_scan.cuh"
+#include <limits>
+
+/// Optional outer namespace(s)
+CUB_NS_PREFIX
+
+/// CUB namespace
+namespace cub
+{
+
+/**
+ * BlockRunLengthDecodeAlgorithm enumerates alternative algorithms for parallel
+ * reduction across a CUDA thread block.
+ */
+enum class BlockRunLengthDecodeAlgorithm
```

Comment:
A few questions bother me:

- If both data and offset types are of the same size, this algorithm requires `2 * BLOCK_ITEMS` shared memory elements for unique items and run offsets and another `2 * BLOCK_ITEMS` for decode buffer. With only two items per thread, the occupancy is already limited by the shared memory size on P100. 

- To get the result, we have to iterate over block tiles. At each iteration, this algorithm has to perform a prefix scan and four thread-block synchronizations. Therefore, algorithm performance is strongly data-dependent. The slowdown should be a multiple of a decoded size divided by the block tile size. 

I've been thinking about an alternative solution. The main idea was that we only need two numbers per thread: value and an offset within the run. I've managed to write [an alternative algorithm ](https://gist.github.com/senior-zero/1d65beb94446cbec80c3c4f336a3f1b0) that requires only `2 * BLOCK_ITEMS` shared memory and doesn't have cooperative iterations after prefix scan computation. It's based on the merge path algorithm and merges prefix sum with the counting iterator. The binary search is performed in shared memory. The approach is similar to the `cub::DeviceSpmv`. The only difference is that instead of row prefix sum, I use runs prefix sums. And instead of a non-zero index, I use the index of an item in the decoded output. 

```cuda
// Prefix sum
int total_decoded_size {};

BlockScanT(temp_storage.scan)
  .InclusiveSum(lengths_reg, lengths_reg, total_decoded_size);
__syncthreads();

const int output_items_per_thread = (TILE_SIZE + total_decoded_size)
                                  / BLOCK_THREADS;

for (int i = 0; i < ITEMS; i++)
{
  int idx = static_cast<int>(threadIdx.x) * ITEMS + i;

  temp_storage.prefix[idx] = lengths_reg[i];
  shared_values[idx] = idx;
}
__syncthreads();

// Merge path
CoordinateT thread_coordinate;

{
  const int diagonal = output_items_per_thread
                     * static_cast<int>(threadIdx.x);

  cub::CountingInputIterator<int> nonzero_indices(0);

  MergePathSearch(diagonal,
                  temp_storage.prefix,
                  nonzero_indices,
                  TILE_SIZE,            // number of unique values (rows)
                  total_decoded_size,   // number of output elements (nnz)
                  thread_coordinate);
}

int value_id = thread_coordinate.x;
int offset = thread_coordinate.y;

int value = shared_values[value_id];
int value_end = temp_storage.prefix[value_id];

for (int i = 0; i < output_items_per_thread; i++)
{
  if (offset >= total_decoded_size)
  {
    break;
  }

  if (offset < value_end)
  {
    CONSUME(value, output); 

    offset++;
  }
  else
  {
    value_id++;
    value = shared_values[value_id];
    value_end = temp_storage.prefix[value_id];
  }
}
```

Preliminary benchmarks show that the speedup is about 3-4x. The speedup depends on the number of decoded items. 

The disadvantage of this approach is that the consecutive threads have items located further from each other. In your implementation, the stride is always equal to the items per thread (which is small). In the merge-path based approach, it's equal to the number of decoded items divided by thread block size, which might be any number.

I think we have two options. 
1) If there's no efficient way to reorder blocked data into the striped arrangement, we can expose this algorithm as a different API. Like `UNUSED`, `OFFSETS` and `NORMAL` implement different algorithms, we could add `MERGE_PATH` one.
2) If there's an efficient way to reorder blocked data into the striped arrangement - we could use only a merge path based approach. 

Both options require some work. I think we should proceed with this PR when the option is chosen.

Reply:
Thanks, I like the idea of using `MergePath` here! 👍  I think we still need to find a solution that will allow to retrieve a coalesced window of items and to iterate over those windows. But that should be doable with a `MergePath`-based approach.
I will also give a simple binary search a try, something like:
```
thread_out_offset = output_items_per_thread * threadIdx.x;
value_id = UpperBound(temp_storage.prefix, num_runs+1, thread_out_offset)-1];
```
That should reduce the binary search from `O(diagonal)` to `O(TILE_SIZE)`.

Reply:
Thanks for the suggestion, @senior-zero! I switched to a binary search-based method. It simplified things quite a bit. No more need to have three different specialisations. We can get all the info we need from the binary search straight.
Would it still be worth to evaluate MergePath? Would we expect a benefit from MergePath over binary search?

Reply:
I'm glad to hear that! 

> Would it still be worth to evaluate MergePath? Would we expect a benefit from MergePath over binary search?
I think that MergePath is a binary search in its essence. If your implementation of binary search doesn't duplicate it and you get similar performance improvements, there's no reason to force MergePath usage 😃